sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnAdapters.java

```diff
@@ -62,6 +62,15 @@ private DoFnAdapters() {}
   @SuppressWarnings({"unchecked", "rawtypes"})
   public static <InputT, OutputT> OldDoFn<InputT, OutputT> toOldDoFn(DoFn<InputT, OutputT> fn) {
     DoFnSignature signature = DoFnSignatures.INSTANCE.getOrParseSignature((Class) fn.getClass());
+
+    if (!signature.stateDeclarations().isEmpty()) {
```

Comment:
I think this would be better done in `ParDo.of(...)` -- it is not required that all runners use `OldDoFn` to run `DoFn`s, so relying on the adapter to catch this problem seems wrong.